### ISSUE TITLE
"resource exhausted (Too many open files)" on a many-core machine

### ISSUE THREAD

**Issue by jwaldmann** [AccelerateHS/accelerate#442]
When I try to run any of accelerate-examples on a machine with many cores, this happens
(not always, but very often):
```
accelerate-nbody --bench --llvm-cpu
...
Warning: Couldn't use randomness source /dev/urandom
Warning: using system clock for seed instead (quality will be lower)
t=0: [Failed]
ERROR: /home/waldmann/.accelerate/accelerate-llvm-1.2.0.1/accelerate-llvm-native-1.2.0.1/llvm-hs-8.0.0/x86_64-unknown-linux-gnu/knl/rel/meep48618995b94e0b9cdcb4c0ac6bea7582d1fcf9dc3b6f20aa82876c6aae0e86cb.o: openBinaryFile: resource exhausted (Too many open files)
...
```

**Comment by tmcdonell** [comment]
Can you give me a bit more information about what your machine, OS, etc.? The initial bug report template lists some of the important information to fill out.

How many is "many cores", for example?

**Comment by jwaldmann** [comment]
5.1.17-300.fc30.x86_64 , "many" = more than 32, ghc-8.6.5

`accelerate` current from hackage (1.2.0.1) but the same happens with the master branch. 

Looking closer at the above output - is this related to https://github.com/AccelerateHS/accelerate-llvm/blob/master/accelerate-llvm/src/Data/Array/Accelerate/LLVM/Compile/Cache.hs ?

When is this called? How often? (Too often, it seems.) Perhaps some finalizers are missing?

**Comment by tmcdonell** [comment]
It should be inspecting the cache once per operation (that is each `map`, `fold` etc.) that it wants to execute, to check whether the code is already compiled. It shouldn't really depend on the number of cores your system has however...

**Comment by tmcdonell** [comment]
I can't reproduce this so closing for now; please get in touch (directly if that works for you) and we can work together to solve this.

**Comment by jwaldmann** [comment]
yes there's probably some other root cause. I just notices that I get the very same error `openFile: resource exhausted (Too many open files)` for some other (non-accelerate) program started with `+RTS -N` (on that machine). If I find the time, I'll look into it and file a GHC RTS issue.

**Comment by robbert-vdh** [comment]
Have you made sure that the open file descriptor limits on your system are not just set really low? You can view the current limit with `ulimit -n`, and you can see how many files a process has opened by counting the files in `/proc/<pid>/fd` (so for instance with Bash you could do `ls -1 /proc/"$(pidof <process_name>)"/fd | wc -l`).
